lib/lookdevXUsd/UsdMxVersionUpgrade.cpp

```diff
@@ -0,0 +1,1097 @@
+    for (const auto& output : ng.GetOutputs())
+    {    
+        for (const auto& sourceInfo : ng.GetConnectedSources(output))
+        {
+            if (sourceInfo.source.GetPrim() == node.GetPrim())
+            {
+                retVal.push_back(std::make_pair(ng, output));
+            }
+        }
+    }
+    return retVal;
+}
+
+_DownstreamInputPortList _getDownstreamInputPorts(UsdShadeConnectableAPI node)
+{
+    // Returns a list(node, port) of all items that connect to the output port of the node passed in as parameter.
+    //    
+    // We assume a nicely behaved graph without connections teleporting across NodeGraph boundaries.
+    _DownstreamInputPortList retVal;
+    auto ng = UsdShadeConnectableAPI(node.GetPrim().GetParent());
+    if (!ng) {
+        return retVal;
+    }
+    // Check every node inside the graph:
+    for (const auto& child : ng.GetPrim().GetChildren())
+    {
+        auto shader = UsdShadeShader(child);
+        if (!shader) {
+            continue;
+        }
+        for (const auto& input : shader.GetInputs()) {
+            for (const auto& sourceInfo : shader.ConnectableAPI().GetConnectedSources(input)) {
+                if (sourceInfo.source.GetPrim() == node.GetPrim()) {
+                    retVal.push_back(std::make_pair(shader.ConnectableAPI(), input));
+                }
+            }
+
+        }
+    }
+
+    return retVal;
+}
+
+void _moveInput(UsdShadeShader sourceNode, const TfToken& sourceInputName, UsdShadeShader destNode, const TfToken& destInputName)
+{
+    auto editor = UsdNamespaceEditor(sourceNode.GetPrim().GetStage());
+    editor.ReparentProperty(sourceNode.GetPrim().GetAttribute(UsdShadeUtils::GetFullName(sourceInputName, UsdShadeAttributeType::Input)),
+                            destNode.GetPrim(), UsdShadeUtils::GetFullName(destInputName, UsdShadeAttributeType::Input));
+    if (editor.CanApplyEdits()) {
+        editor.ApplyEdits();
+    } else {
+        TF_WARN("Failed to move input '%s' from node '%s' to input '%s' on node '%s'. Please make sure the material layer is writable.",
+                sourceInputName.GetText(),
+                sourceNode.GetPrim().GetPath().GetText(),
+                destInputName.GetText(),
+                destNode.GetPrim().GetPath().GetText());
+    }
+}
+
+void _upgradeMaterial(UsdShadeMaterial usdMaterial)
+{
+    TF_AXIOM(usdMaterial);
+
+    // This is the upgrade from 1.38 to 1.39. Feel free to split into multiple
+    // functions if the upgrade process gets more complex.
+
+    // If this material is already 1.39, then nothing to do:
+    if (usdMaterial.GetPrim().HasAPI<UsdMtlxMaterialXConfigAPI>()) {
+        auto configAPI = UsdMtlxMaterialXConfigAPI(usdMaterial.GetPrim());
+        auto versionAttr = configAPI.GetConfigMtlxVersionAttr();
+        std::string versionStr;
+        if (versionAttr && versionAttr.Get(&versionStr)) {
+            if (versionStr == _tokens->currentMxVersion.GetString()) {
```

Comment:
For the future: will we want to upgrade when MaterialX is 1.40 for example? When we are at 1.40, this will make it so the code will try to upgrade 1.39 materials. I guess it will be a no-op anyway, no harm.

Reply:
The MaterialX strategy is to upgrade in sequence, so if we get to 1.41 and load a legacy document still at 1.38, then it has to run thru the 1.38->1.39->1.40->1.41 gauntlet.

MaterialX chose to put all code blocks into a single "upgrade" function, but the MayaUSD version should definitely split them by renaming this function `_upgradeMaterialTo1_39` and create a global `_upgradeMaterial` function that checks the current version and starts the gauntlet at the right step. 

Makes this check unnecessary (and it was wrong to start with as it should have been >= 1.39 to be future-proof.